src/parser/core/SharedCode.tsx

```diff
@@ -0,0 +1,231 @@
+    return this.getEvents(true, EventType.Cast, preCastSpell, 1, event.timestamp, 250)[0];
+  }
+
+  /**
+   * @param eventType The type of event you want to count. i.e. EventType.Cast, EventType.BeginCast, etc. Use EventType.Event for all events.
+   * @param spell the spell you want to count casts for.
+   * @returns the number of events found
+   */
+  countEvents(eventType: EventType, spell: SpellInfo) {
+    const events = this.getEvents(true, eventType, spell);
+    return events.length;
+  }
+
+  /**
+   * @param buffActive filters based on whether the buff is active or inactive (true for active, false for inactive)
+   * @param buff the spell object for the buff
+   * @param eventType the type of event that you want to search for. i.e. EventType.Cast, EventType.BeginCast, etc. Use EventType.Event for all events.
+   * @param cast an optional cast spell object to count. Omit or leave undefined to count all casts
+   * @returns the number of events found
+   */
+  countEventsByBuff(buffActive: boolean, buff: SpellInfo, eventType: EventType, cast?: SpellInfo) {
+    const events = buffActive
+      ? this.getEventsByBuff(true, buff, eventType, cast)
+      : this.getEventsByBuff(false, buff, eventType, cast);
+    return events.length;
+  }
+
+  /**
+   * @param buffActive filters based on whether the buff is active or inactive (true for active, false for inactive)
+   * @param buff the spell object for the buff
+   * @param eventType the type of event that you want to search for. i.e. EventType.Cast, EventType.BeginCast, etc. Use EventType.Event for all events.
+   * @param spell an optional spell object to search. Omit or leave undefined to count all events
+   * @returns an array of events that match the provided criteria
+   */
+  getEventsByBuff<ET extends EventType>(
+    buffActive: boolean,
+    buff: SpellInfo,
+    eventType: ET,
+    spell?: SpellInfo | SpellInfo[],
+  ): Array<MappedEvent<ET>> {
+    const events = this.getEvents(true, eventType, spell);
+    const filteredEvents = events.filter((e) =>
+      buffActive
+        ? this.selectedCombatant.hasBuff(buff.id, e.timestamp - 1)
+        : !this.selectedCombatant.hasBuff(buff.id, e.timestamp + 1),
+    );
+    return filteredEvents;
+  }
+
+  /**
+   * @param buff the spell object for the proc's buff.
+   * @param spenderSpell the spell object (or an array of spell objects) that are used to spend the proc.
+   * @returns an array of remove buff events that had expired
+   */
+  getExpiredProcs(buff: SpellInfo, spenderSpell: SpellInfo | SpellInfo[]) {
+    const events = this.getEvents(true, EventType.RemoveBuff, buff);
+
+    const filteredEvents = events.filter((e) => {
+      const castEvent = this.getEvents(
+        true,
+        EventType.Cast,
+        spenderSpell,
+        1,
+        e.timestamp + 1,
+        250,
+      )[0];
+      return !castEvent;
+    });
+    return filteredEvents;
+  }
```

Comment:
this implementation is a little bit spooky to me because of how it could theoretically blow up by calling `getEvents` for every event, but I can't really think of a realistic scenario where that causes problems.